sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java

```diff
@@ -595,58 +608,65 @@ public BasicDocumentRevision call(SQLDatabase db) throws Exception {
+        // check if the docid exists first:
 
-            // if it does exist:
-            // * if winning leaf deleted, root the 'created' document there
-            // * else raise error
-            // if it does not exist:
-            // * normal insert logic for a new document
+        // if it does exist:
+        // * if winning leaf deleted, root the 'created' document there
+        // * else raise error
+        // if it does not exist:
+        // * normal insert logic for a new document
 
-            InsertRevisionOptions options = new InsertRevisionOptions();
-            BasicDocumentRevision potentialParent = this.getDocumentInQueue(db, docId, null);
-            if (potentialParent != null) {
-                if (!potentialParent.isDeleted()) {
-                    // current winner not deleted, can't insert
-                    throw new IllegalArgumentException("Can not insert new doc, likely the docId exists already: "
-                            + docId);
-                }
-                // if we got here, parent rev was deleted
-                this.setCurrent(db,potentialParent, false);
-                options.revId = CouchUtils.generateNextRevisionId(potentialParent.getRevision());
-                options.docNumericId = potentialParent.getInternalNumericId();
-                options.parentSequence = potentialParent.getSequence();
-            } else {
-                // otherwise we are doing a 'normal' create document
-                long docNumericId = insertDocumentID(db,docId);
-                options.revId = CouchUtils.getFirstRevisionId();
-                options.docNumericId = docNumericId;
-                options.parentSequence = -1l;
-            }
+        InsertRevisionOptions options = new InsertRevisionOptions();
+        BasicDocumentRevision potentialParent = null;
 
-            options.deleted = false;
-            options.current = true;
-            options.data = body.asBytes();
-            options.available = true;
-            options.copyAttachments = false;
-            long newSequence = insertRevision(db,options);
-            if (newSequence < 0) {
-                throw new IllegalStateException("Error inserting new revision: "+options);
-            }
+        try {
+            potentialParent = this.getDocumentInQueue(db, docId, null);
+        } catch (DocumentNotFoundException e) {
+            //this is an expected exception, it just means we are
+            // resurrecting the document
+        }
 
-            BasicDocumentRevision doc = getDocumentInQueue(db, docId, options.revId);
-            documentCreated = new DocumentCreated(doc);
+        if (potentialParent != null) {
+            if (!potentialParent.isDeleted()) {
+                // current winner not deleted, can't insert
+                throw new IllegalArgumentException("Can not insert new doc, likely the docId exists already: "
+                        + docId);
```

Comment:
Does this get thrown as a ConflictException somewhere else; it sure sounds like one.